src/sipo.rs

```diff
@@ -38,109 +49,103 @@ impl OutputPin for ShiftRegisterPin<'_>
     }
 }
 
-macro_rules! ShiftRegisterBuilder {
-    ($name: ident, $size: expr) => {
-        /// Serial-in parallel-out shift register
-        pub struct $name<Pin1, Pin2, Pin3>
-            where Pin1: OutputPin,
-                  Pin2: OutputPin,
-                  Pin3: OutputPin
-        {
-            clock: RefCell<Pin1>,
-            latch: RefCell<Pin2>,
-            data: RefCell<Pin3>,
-            output_state: RefCell<[bool; $size]>,
-        }
-
-        impl<Pin1, Pin2, Pin3> ShiftRegisterInternal for $name<Pin1, Pin2, Pin3>
-            where Pin1: OutputPin,
-                  Pin2: OutputPin,
-                  Pin3: OutputPin
-        {
-            /// Sets the value of the shift register output at `index` to value `command`
-            fn update(&self, index: usize, command: bool) -> Result<(), ()>{
-                self.output_state.borrow_mut()[index] = command;
-                let output_state = self.output_state.borrow();
-                self.latch.borrow_mut().set_low().map_err(|_e| ())?;
-
-                for i in 1..=output_state.len() {
-                    if output_state[output_state.len() - i] {
-                        self.data.borrow_mut().set_high().map_err(|_e| ())?;
-                    } else {
-                        self.data.borrow_mut().set_low().map_err(|_e| ())?;
-                    }
-                    self.clock.borrow_mut().set_high().map_err(|_e| ())?;
-                    self.clock.borrow_mut().set_low().map_err(|_e| ())?;
-                }
+/// Serial-in parallel-out shift register
+pub struct ShiftRegister<Pin1, Pin2, Pin3, const N: usize>
+where
+    Pin1: OutputPin,
+    Pin2: OutputPin,
+    Pin3: OutputPin,
+{
+    clock: RefCell<Pin1>,
+    latch: RefCell<Pin2>,
+    data: RefCell<Pin3>,
+    output_state: RefCell<[bool; N]>,
+}
 
-                self.latch.borrow_mut().set_high().map_err(|_e| ())?;
-                Ok(())
-            }
+impl<Pin1, Pin2, Pin3, const N: usize> ShiftRegister<Pin1, Pin2, Pin3, N>
+where
+    Pin1: OutputPin,
+    Pin2: OutputPin,
+    Pin3: OutputPin,
+{
+    /// Creates a new SIPO shift register from clock, latch, and data output pins
+    pub fn new(clock: Pin1, latch: Pin2, data: Pin3) -> Self {
+        ShiftRegister {
```

Comment:
```suggestion
        assert!(N > 0 && N % 8 == 0);
        ShiftRegister {
```

Reply:
I also think that.
All shift registers are 8bit or multiple of 8-bit?
Yes, I also think all are!!:)